### PR TITLE
[WIP] Add Hadoop configuration files to driver pod.

### LINKED ISSUE #130: Ship Hadoop configuration files to the driver and add to its classpath
Currently we do not do this, so the only way Hadoop configuration options can be set is by setting `spark.hadoop.*` parameters on the Spark configuration.

### PR DESCRIPTION
**Pull request by mccheah**
Closes #130

### CODE REVIEW

**Comment by mccheah** [comment]
The idea is here but missing adding the configuration files to the executors as well as tests.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        val isNull = files._2 == null
+        if (isNull) {
+          logWarning(s"Failed to list files from $confEnvVar at ${files._1.getAbsolutePath}")
+        }
+        isNull
+      }.flatMap(_._2.toSeq)
+      .filter { file =>
+        val isFile = file.isFile
+        // We could theoretically make a way to load directories, but this isn't supported by YARN
+        // and simplifies the logic greatly. Might be worth considering in a future iteration.
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala:76]: I think adding `/etc/hadoop/conf` to the classpath makes `new Configuration(true)` load defaults from files in that directory.  Would have to verify that

**Comment by ash211** [comment]
@kimoonkim do you often see credentials in your `*-site.xml` files?  We've tried pretty hard to eliminate sensitive material from those files over here and don't have any remaining, so aren't sure how sensitive to be treating these files in general.

@foxish are k8s config maps a safe enough way to send these files to the driver pod (pre-launch) or, if these contain sensitive materials, should we be using secrets instead?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    )
+  }
+
+  private def getMountedHadoopConfiguration(
+      confEnvVar: String,
+      configMapName: String,
+      mountPath: String): MountedHadoopConfiguration = {
+    val confFilesContents: Map[String, String] = sys.env.get(confEnvVar)
+      .map(new File(_))
+      .filter(_.isDirectory)
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala:64]: Would you mind popping an INFO log line in here somewhere saying something along the lines of `Sending the contents of /path/to/dir as HADOOP_CONF_DIR to the driver`

**Comment by kimoonkim** [comment]
> @kimoonkim do you often see credentials in your *-site.xml files? We've tried pretty hard to eliminate sensitive material from those files over here and don't have any remaining, so aren't sure how sensitive to be treating these files in general.

@ash211 The majority of the time, YARN uses them for storing *cluster-wide information*, and I think that's the context here as well. If yes, I don't think we see a lot of sensitive information in these files.  (@ssuchter would know more)  Some uneducated users might mistake these for storing *job specific information* which could be potentially sensitive, but I am not sure what we can do about that.

**Comment by ssuchter** [comment]
I agree. The *-site.xml files generally don’t have sensitive information.
Sometimes they contain pointers to other files (that have more restricted
permissions) that contain sensitive information. However, that use case
doesn’t really apply to us here, because of the difference between
momentary dockers vs long-lived hosts.

The job configs sometimes have semi-sensitive information, and sometimes
credentials, but good sites avoid that.

**Comment by ash211** [comment]
Based on that I think it's safe to defer on confidentially transferring these files to the pods -- in practical terms config maps probably the right choice here, and k8s secrets or more intricate mechanisms not worth the effort.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+private[spark] class MountedHadoopConfigurationsProvider(kubernetesAppId: String)
+    extends Logging {
+
+  def get(): MountedHadoopConfigurations = {
+    MountedHadoopConfigurations(
+      getMountedHadoopConfiguration(
+        confEnvVar = "HADOOP_CONF_DIR",
+        configMapName = s"$kubernetesAppId-hadoopConfDir",
+        mountPath = "/etc/hadoop/hadoopConfDir"),
```
Comment (lins05) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala:50]: Why not `/etc/hadoop/conf` ? I think it's the default hadoop conf dir for most hadoop distros.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      .map(new File(_))
+      .filter(_.isDirectory)
+      .map(dir => (dir, dir.listFiles()))
+      .filterNot { files =>
+        val isNull = files._2 == null
+        if (isNull) {
+          logWarning(s"Failed to list files from $confEnvVar at ${files._1.getAbsolutePath}")
+        }
+        isNull
+      }.flatMap(_._2.toSeq)
```
Comment (lins05) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/MountedHadoopConfigurationsProvider.scala:72]: Got a scala compilation error here.

**Comment by foxish** [comment]
> @foxish are k8s config maps a safe enough way to send these files to the driver pod (pre-launch) or, if these contain sensitive materials, should we be using secrets instead?

Neither of them is opaque really, but I'd go with secrets and not configmaps for credentials.

**Comment by mccheah** [comment]
Will re-do this on top of submission v2.

**Pull request closed without merging**